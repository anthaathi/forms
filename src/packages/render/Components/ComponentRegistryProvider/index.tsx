import * as React from 'react';
import { useEffect } from 'react';
import {
  ComponentRegistry,
  ComponentRegistryProps,
} from '../../Context/ComponentRegistry';
import validateComponentRegistryElements from './utils/validate-component-registry-elements';

export interface ComponentRegistryProviderProps {
  elements: ComponentRegistryProps[];
  children: React.ReactNode;
}

/**
 * Provides and validate the elements
 * @param elements
 * @param children
 * @constructor
 */
export default function ComponentRegistryProvider({
  elements,
  children,
}: ComponentRegistryProviderProps) {
  useEffect(() => {
    const isValid = validateComponentRegistryElements(elements);

    if (!isValid) {
      // eslint-disable-next-line no-console
      console.warn('all element types should start with capital name');
    }
  }, [elements]);

  return (
    <ComponentRegistry.Provider value={elements}>
      {children}
    </ComponentRegistry.Provider>
  );
}
