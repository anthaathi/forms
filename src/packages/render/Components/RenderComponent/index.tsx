import * as React from 'react';
import { useEffect, useMemo } from 'react';
import { useComponentRegistry } from '../../Context/ComponentRegistry';
import startWithCapital from './utils/start-with-capital';

export interface RenderComponentsProps<T> {
  schema: {
    $$kind: string;
    children?: React.ReactNode;
  } & T;

  version: string;
}

/**
 * Render dynamic components just like json schema
 * @constructor
 */
export default function RenderComponent<T = any>({
  schema: { $$kind, ...props },
  version,
}: RenderComponentsProps<T>) {
  const componentRegistry = useComponentRegistry();

  const component = useMemo(
    () => componentRegistry.find((res) => res.type === $$kind),
    [$$kind, componentRegistry],
  );

  const propsToPass = useMemo(() => {
    const newProps = {
      ...props,
    } as T & any;

    function convertApp(value: any, index: number) {
      if (React.isValidElement(value)) {
        return value;
      }

      if (value && typeof value === 'object' && '$$kind' in value) {
        return (
          <RenderComponent
            key={value.id || index}
            version={version}
            schema={value}
          />
        );
      }

      if (!value) {
        return null;
      }

      return value;
    }

    Object.keys(newProps).forEach((res) => {
      if (res.startsWith('$$')) {
        delete newProps[res];
        return;
      }

      const value = newProps[res];

      if (Array.isArray(value)) {
        newProps[res] = value.map(convertApp);
      } else {
        newProps[res] = convertApp(value, 0);
      }
    });

    return newProps;
  }, [props, version]);

  const shouldRender = useMemo(
    () => component?.component || !startWithCapital($$kind),
    [$$kind, component?.component],
  );

  if (process.env.NODE_ENV !== 'production') {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    useEffect(() => {
      if (!shouldRender) {
        // eslint-disable-next-line no-console
        console.warn(`${$$kind} is not registered`);
      }
    }, [$$kind, shouldRender]);
  }

  if (!shouldRender) {
    return null;
  }

  return React.createElement(
    component?.component || $$kind,
    propsToPass,
    propsToPass?.children || null,
  );
}
