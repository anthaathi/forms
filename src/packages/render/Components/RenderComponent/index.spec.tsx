/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { render } from '@testing-library/react';
import { Block } from 'baseui/block';
import { BaseProvider, LightTheme } from 'baseui';
import { Provider as StyletronProvider } from 'styletron-react';
import { Client as Styletron } from 'styletron-engine-atomic';
import RenderComponent from './index';
import ComponentRegistryProvider from '../ComponentRegistryProvider';

describe('RenderComponent', () => {
  it('should render components', () => {
    const consoleWarnMock = jest
      .spyOn(global.console, 'warn')
      .mockImplementation();
    const tree = render(
      <RenderComponent
        version="alpha1"
        schema={{
          $$kind: 'div',
          $$version: 'alpha0',
          'data-id': 'item-rendered1',
          children: [
            {
              $$kind: 'div',
              'data-id': 'item-rendered2',
              children: <div>Hello world</div>,
            },
            {
              $$kind: 'div',
              'data-id': 'item-rendered3',
              children: null,
            },
            {
              $$kind: 'Div',
              'data-id': 'item-rendered4',
              children: null,
            },
          ],
        }}
      />,
    );

    expect(global.console.warn).toHaveBeenCalledTimes(1);

    consoleWarnMock.mockRestore();

    expect(tree.container).toMatchSnapshot();
  });

  it('should render tree with passed components', () => {
    const engine = new Styletron({
      prefix: '_',
    });

    const tree = render(
      <StyletronProvider value={engine}>
        <BaseProvider theme={LightTheme}>
          <ComponentRegistryProvider
            elements={[
              {
                type: 'Block',
                component: Block,
              },
            ]}
          >
            <RenderComponent
              version="alpha0"
              schema={{
                $$kind: 'div',
                children: {
                  $$kind: 'Block',

                  $style: {
                    marginTop: '12px',
                  },
                  children: [
                    {
                      $$kind: 'span',
                      children: 'test',
                    },
                    {
                      $$kind: 'span',
                      children: 'test2',
                    },
                  ],
                },
              }}
            />
          </ComponentRegistryProvider>
        </BaseProvider>
      </StyletronProvider>,
    );

    expect(tree.container).toMatchSnapshot();
  });
});
