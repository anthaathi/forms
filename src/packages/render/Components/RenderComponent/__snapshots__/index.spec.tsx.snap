// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderComponent should render components 1`] = `
<div>
  <div
    data-id="item-rendered1"
  >
    <div
      data-id="item-rendered2"
    >
      <div>
        Hello world
      </div>
    </div>
    <div
      data-id="item-rendered3"
    />
  </div>
</div>
`;

exports[`RenderComponent should render tree with passed components 1`] = `
<div>
  <div
    class=""
  >
    <div>
      <div
        class="_ae"
        data-baseweb="block"
      >
        <span>
          test
        </span>
        <span>
          test2
        </span>
      </div>
    </div>
  </div>
  <div
    class=""
  />
</div>
`;
